services/src/projects/project.rs

```diff
@@ -0,0 +1,178 @@
+use uuid::Uuid;
+use crate::workflows::Workflow;
+use crate::users::user::{UserIdentification, UserInput};
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+use crate::error::Error;
+use crate::error;
+use snafu::ensure;
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Copy, Hash)]
+pub struct ProjectId {
+    id: Uuid
+}
+
+impl ProjectId {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            id: Uuid::new_v4()
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub struct Project {
+    pub id: ProjectId,
+    pub version: usize,
+    pub changed: DateTime<Utc>,
+    pub name: String,
+    pub description: String,
+    pub layers: Vec<Layer>,
+    //TODO: stref
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub struct Layer {
+    pub workflow: Workflow,
+    pub name: String,
+    //TODO: colorizer
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub enum ProjectOwner {
+    User { user: UserIdentification },
+    Any,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub enum OrderBy {
+    DateAsc,
+    DateDesc,
+    NameAsc,
+    NameDesc
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub struct ProjectListing {
+    pub id: ProjectId,
+    pub name: String,
+    pub description: String,
+    pub layer_name: Vec<String>,
+    pub changed: DateTime<Utc>
+}
+
+impl From<Project> for ProjectListing {
+    fn from(project: Project) -> Self {
+        Self {
+            id: project.id,
+            name: project.name,
+            description: project.description,
+            layer_name: project.layers.into_iter().map(|l| l.name).collect(),
+            changed: project.changed
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub enum ProjectFilter {
+    Name { term: String },
+    Description { term: String },
+    None,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone, Hash)]
+pub struct CreateProject {
+    pub name: String,
+    pub description: String,
+    // TODO: stref
+}
+
+impl UserInput for CreateProject {
+    fn validate(&self) -> Result<(), Error> {
```

Comment:
We have to be kind of careful because the fields of `CreateProject` are `pub`, so you could change them from outside after you called validate(), so a Validated<CreateProject> could become invalid.

This could be tedious to change, but we have to consider pros and cons at least.

Reply:
once it's validated, the user input should be moved into a Validated<UserInput> object. is there a way to make this read only then?

Reply:
Yes, you must never allow mutable access to Validated members, so only `&self` access.

Reply:
so leave it like it is for now?

Reply:
Yes